File: cplex/examples/src/matlab/miqpex1.m
function miqpex1()
% Enter and optimize a mixed integer quadratic programming problem
%
% This function fills in the data structures for the mixed integer
% quadratic program:
%
%       Maximize
%        obj: x1 + 2 x2 + 3 x3 + x4
%             - 0.5 ( 33x1*x1 + 22*x2*x2 + 11*x3*x3
%                    -  12*x1*x2 - 23*x2*x3 )
%       Subject To
%        c1: - x1 + x2 + x3 + 10x4  <= 20
%        c2: x1 - 3 x2 + x3         <= 30
%        c3:       x2       - 3.5x4  = 0
%       Bounds
%        0 <= x1 <= 40
%        0 <= x2
%        0 <= x3
%        0 <= x4 <= 3
%       Integers
%         x4
%       End

% ---------------------------------------------------------------------------
% File: miqpex1.m
% Version 12.6.1
% ---------------------------------------------------------------------------
% Licensed Materials - Property of IBM
% 5725-A06 5725-A29 5724-Y48 5724-Y49 5724-Y54 5724-Y55 5655-Y21
% Copyright IBM Corporation 2008, 2014. All Rights Reserved.
%
% US Government Users Restricted Rights - Use, duplication or
% disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
% ---------------------------------------------------------------------------

try
   % Initialize the CPLEX object
   cplex = Cplex('miqpex1');
   
   % Fill in the data for the problem using populatebyrow
   populatebyrow (cplex);
   
   % Optimize the problem
   cplex.solve();
   
   % Write the solution
   fprintf ('\nSolution status = %s \n', cplex.Solution.statusstring);
   fprintf ('Solution value = %f \n', cplex.Solution.objval);
   disp ('Values =');
   disp (cplex.Solution.x);
   disp ('Slacks =');
   disp (cplex.Model.rhs - cplex.Solution.ax);
catch m
   disp(m.message);
end
   function populatebyrow(cplex)
      cplex.Model.sense = 'maximize';
      cplex.addCols([1 2 3 1]', [], [0 0 0 0]', [40 inf  inf 3]', 'CCCI');
      cplex.addRows(-inf, [-1 1 1 10], 20);
      cplex.addRows(-inf, [1 -3 1  0], 30);
      cplex.addRows(0, [0 1 0 -3.5], 0);
      cplex.Model.Q = [-33     6     0  0;
         6   -22  11.5  0;
         0  11.5   -11  0;
         0     0     0  0];
   end
end

File: cplex/examples/src/java/MIQPex1.java
/* --------------------------------------------------------------------------
 * File: MIQPex1.java
 * Version 12.6.1
 * --------------------------------------------------------------------------
 * Licensed Materials - Property of IBM
 * 5725-A06 5725-A29 5724-Y48 5724-Y49 5724-Y54 5724-Y55 5655-Y21
 * Copyright IBM Corporation 2001, 2014. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or
 * disclosure restricted by GSA ADP Schedule Contract with
 * IBM Corp.
 * --------------------------------------------------------------------------
 *
 * MIQPex1.java - Entering and optimizing a MIQP problem
 */

import ilog.concert.*;
import ilog.cplex.*;


public class MIQPex1 {
   public static void main(String[] args) {
      try {
         IloCplex cplex = new IloCplex();
         IloLPMatrix lp = populateByRow(cplex);

         cplex.setParam(IloCplex.Param.RootAlgorithm, IloCplex.Algorithm.Dual);


         if ( cplex.solve() ) {
            double[] x     = cplex.getValues(lp);
            double[] slack = cplex.getSlacks(lp);

            System.out.println("Solution status = " + cplex.getStatus());
            System.out.println("Solution value  = " + cplex.getObjValue());

            int nvars = x.length;
            for (int j = 0; j < nvars; ++j) {
               System.out.println("Variable " + j +
                                  ": Value = " + x[j]);
            }

            int ncons = slack.length;
            for (int i = 0; i < ncons; ++i) {
               System.out.println("Constraint " + i +
                                  ": Slack = " + slack[i]);
            }

            cplex.exportModel("miqpex1.lp");
         }
         cplex.end();
      }
      catch (IloException e) {
         System.err.println("Concert exception '" + e + "' caught");
      }
   }

   static IloLPMatrix populateByRow(IloMPModeler model) throws IloException {
      IloLPMatrix lp = model.addLPMatrix();

      double[]    lb = {0.0, 0.0, 0.0};
      double[]    ub = {40.0, Double.MAX_VALUE, Double.MAX_VALUE};
      IloNumVar[] x  = model.numVarArray(model.columnArray(lp, 3), lb, ub);
      IloNumVar   y  = model.intVar(model.column(lp), 0, 3);

      // - x0 +   x1 + x2 + 10*y <= 20
      //   x0 - 3*x1 + x2        <= 30
      double[]   lhs = {-Double.MAX_VALUE, -Double.MAX_VALUE};
      double[]   rhs = {20.0, 30.0};
      double[][] val = { {-1.0,  1.0,  1.0, 10.0},
                         { 1.0, -3.0,  1.0} };
      int[][]    ind = { {0, 1, 2, 3},
                         {0, 1, 2} };
      lp.addRows(lhs, rhs, ind, val);
      // x1 - 3.5*y = 0
      lp.addRow(model.eq(model.diff(x[1], model.prod(3.5, y)), 0.0));

      // Q = 0.5 ( 33*x0*x0 + 22*x1*x1 + 11*x2*x2 - 12*x0*x1 - 23*x1*x2 )
      IloNumExpr x00 = model.prod( 33, model.square(x[0]));
      IloNumExpr x11 = model.prod( 22, model.square(x[1]));
      IloNumExpr x22 = model.prod( 11, model.square(x[2]));
      IloNumExpr x01 = model.prod(-12, model.prod(x[0], x[1]));
      IloNumExpr x12 = model.prod(-23, model.prod(x[1], x[2]));
      IloNumExpr Q   = model.prod(0.5, model.sum(x00, x11, x22, x01, x12));

      // maximize x0 + 2*x1 + 3*x2 + Q
      double[] objvals = {1.0, 2.0, 3.0};
      model.add(model.maximize(model.diff(model.scalProd(x, objvals), Q)));

      return (lp);
   }
}
